### PR TITLE
(CDAP-15106) Fix TMS overlapping lock issue with LevelDB

### PR DESCRIPTION
**Pull request by chtyim**

### CODE REVIEW

**Comment by chtyim** [comment]
Build passed https://builds.cask.co/browse/CDAP-DUT6938

File: cdap-tms/src/main/java/co/cask/cdap/messaging/store/leveldb/LevelDBTableFactory.java [hunk 1]
```diff
@@ -201,18 +200,21 @@ public void run() {
...
-      }
 
-      long timeStamp = System.currentTimeMillis();
-      try {
-        for (Map.Entry<TopicMetadata, LevelDBMessageTable> messageTableEntry : messageTableCache.asMap().entrySet()) {
-          messageTableEntry.getValue().pruneMessages(messageTableEntry.getKey(), timeStamp);
-        }
-        for (Map.Entry<TopicMetadata, LevelDBPayloadTable> payloadTableEntry : payloadTableCache.asMap().entrySet()) {
-          payloadTableEntry.getValue().pruneMessages(payloadTableEntry.getKey(), timeStamp);
+          // Prune the current generation
```
Comment (anew) [inline cdap-tms/src/main/java/co/cask/cdap/messaging/store/leveldb/LevelDBTableFactory.java:204]: Not sure I understand this correctly: the old code was pruning all the files in the cache (is that, all generations?). Now that the cache has been refactored, why do only prune the current generation now?

**Review by anew — COMMENTED** [review]
Only had one question

File: cdap-tms/src/main/java/co/cask/cdap/messaging/store/leveldb/LevelDBTableFactory.java (hunk 1 above)
Comment (chtyim) [reply cdap-tms/src/main/java/co/cask/cdap/messaging/store/leveldb/LevelDBTableFactory.java:204]: In the new code, it first cleanup all older generations by closing the LevelDB, removing the cached object as well as deleting file path (line 178-201). Then it prunes current generation if the levelDB object is in the cache (meaning someone is using it). The cleanup of the current generation won't happen if no one is using the topic in the current JVM.

Comment (anew) [reply cdap-tms/src/main/java/co/cask/cdap/messaging/store/leveldb/LevelDBTableFactory.java:204]: Oh I see, that means there is always only one generation? I thought there can be generation older than the current one, that still cannot be deleted, but should it be pruned? 

Just trying to understand the effect of this change

Comment (anew) [reply cdap-tms/src/main/java/co/cask/cdap/messaging/store/leveldb/LevelDBTableFactory.java:204]: Git it now. LGTM

**Review by anew — APPROVED** [review]

**Pull request merged**